trustm3/device_fraunhofer_common_cml
# Feature/usbtoken

This PR aims to enable the usage of hardware token to wrap the FDE keys for container data encryption.
It relies on the availability of a ctccid and a cardservice library from CardContact. Both of which can be build from source from [1]. To facilitate the testing, a fork of meta-trustx layer can be checked out to your local workspace from [2] which already includes the necessary changes.

If trustm3 is built with the adapted meta-trustx layer the cml layer includes the "key-generator" example as described in [3].

To test the key wrapping using the usbtoken the following procedure can be applied:
1. Create a container config that includes usage of a usbtoken. It may look like this:
     ```
    name: "usbtoken_test"
    guest_os: "trustx-coreos"
    guestos_version: 20200522073051
    type: CONTAINER
    token_type: USB
    usb_configs {
        id: "0x04e6:0x5816"
        serial: "55511725600361"
        assign: false
        type: TOKEN
    }
    ```

2. Ensure the token is initialized (this operation sets the transport PIN and is only possible if the user can present the valid SO PIN). Invoke the following command from the CML debugging shell:
    ```bash
    key-generator --init --transportpin 123456
    ```
    For demo purpose, a valid SO PIN for the development token is built into the "key-generator".

3. Create the container
    ```bash
    control create usbtoken_test.skel
    ```

4. The token must be paired to the device before it can be used. The pairing is automatically done on first invocation of ```control change_pin```. Therefore, invoke
    ```bash
    control change_pin usbtoken_test
    ```
   and enter the transport PIN and the new User PIN when prompted.

5. Start the container 

TODOs:
- [ ] The pairing secret should be stored securely in a platform bound manner, e.g. in a TPM

[1] https://github.com/CardContact/sc-hsm-embedded
[2] https://github.com/ceppleaisec/meta-trustx
[3] https://github.com/CardContact/sc-hsm-embedded/tree/master/src/examples/key-generator
